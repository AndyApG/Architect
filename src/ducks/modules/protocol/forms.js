const initialState = {
  person: {
    title: 'Add A Person',
    entity: 'node',
    type: 'person',
    fields: [
      {
        variable: 'name',
        component: 'TextInput',
      },
      {
        variable: 'nickname',
        component: 'TextInput',
      },
      {
        variable: 'age',
        component: 'TextInput',
      },
      {
        variable: 'timeCreated',
        component: 'hidden',
        value: 'return Date.now().toString();',
      },
    ],
  },
};

export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    default:
      return state;
  }
}

const actionCreators = {
};

const actionTypes = {
};

export {
  actionCreators,
  actionTypes,
};
