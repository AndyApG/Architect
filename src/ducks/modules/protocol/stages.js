import uuid from 'uuid/v1';
import { get } from 'lodash';

const CREATE_STAGE = Symbol('PROTOCOL/CREATE_STAGE');
const UPDATE_STAGE = Symbol('PROTOCOL/UPDATE_STAGE');
const DELETE_STAGE = Symbol('PROTOCOL/DELETE_STAGE');

const initialState = [];
const initialStage = {
  label: '',
};

export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case CREATE_STAGE: {
      const stage = { ...initialStage, ...action.stage, id: uuid() };
      const insertAtIndex = get(action, 'index', state.length);

      return [
        ...state.slice(0, insertAtIndex),
        stage,
        ...state.slice(insertAtIndex),
      ];
    }
    case UPDATE_STAGE:
      return state.map((stage) => {
        if (stage.id !== action.id) { return stage; }

        if (action.overwrite) { return { ...action.stage, id: stage.id }; }

        return {
          ...stage,
          ...action.stage,
          id: stage.id,
        };
      });
    case DELETE_STAGE:
      return state.filter(stage => (stage.id !== action.id));
    default:
      return state;
  }
}

function createStage(stage, index) {
  return {
    type: CREATE_STAGE,
    stage,
    index,
  };
}

function updateStage(stageId, stage, overwrite = false) {
  return {
    type: UPDATE_STAGE,
    id: stageId,
    stage,
    overwrite,
  };
}

function deleteStage(stageId) {
  return {
    type: DELETE_STAGE,
    id: stageId,
  };
}

const actionCreators = {
  createStage,
  updateStage,
  deleteStage,
};

const actionTypes = {
  CREATE_STAGE,
  UPDATE_STAGE,
  DELETE_STAGE,
};

export {
  actionCreators,
  actionTypes,
};
