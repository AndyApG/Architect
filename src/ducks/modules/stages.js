import { map, max } from 'lodash';

const ADD_STAGE = 'ADD_STAGE';

const initialState = [{ id: 1, type: 'NameGenerator' }, { id: 2, type: 'Sociogram' }];

const nextId = state => (max(map(state, 'id')) + 1);

export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case ADD_STAGE:
      return [
        ...state.slice(0, action.index),
        { ...action.options, id: nextId(state) },
        ...state.slice(action.index),
      ];
    default:
      return state;
  }
}

function addStage(options, index) {
  return {
    type: ADD_STAGE,
    options,
    index,
  };
}

const actionCreators = {
  addStage,
};

const actionTypes = {
  ADD_STAGE,
};

export {
  actionCreators,
  actionTypes,
};
