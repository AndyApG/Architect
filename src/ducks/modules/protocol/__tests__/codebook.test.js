/* eslint-env jest */

import uuid from 'uuid';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { sortBy } from 'lodash/fp';
import { actionCreators as formActions } from '../forms';
import { actionCreators as stageActions } from '../stages';
import reducer, { actionTypes, actionCreators, testing } from '../codebook';

jest.mock('uuid');

const sortByType = sortBy('type');

const mockStore = configureStore([thunk]);

const mockState = {
  node: {
    place: { foo: 'bar' },
    person: { hello: 'world' },
  },
  edge: { },
};

describe('protocol.codebook', () => {
  describe('reducer', () => {
    it('initialState', () => {
      const initialState = reducer();

      expect(initialState).toEqual({
        node: {},
        edge: {},
      });
    });

    it('CREATE_TYPE', () => {
      const result = reducer(
        undefined,
        testing.createType('node', 'foo', { bar: 'bazz' }),
      );

      expect(result).toEqual({
        node: { foo: { bar: 'bazz' } },
        edge: {},
      });
    });

    it('UPDATE_TYPE', () => {
      const result = reducer(
        {
          node: { foo: { bar: 'bazz' } },
          edge: {},
        },
        actionCreators.updateType('node', 'foo', { fizz: 'pop' }),
      );

      expect(result).toEqual({
        node: { foo: { fizz: 'pop' } },
        edge: {},
      });
    });

    it('DELETE_TYPE', () => {
      const result = reducer(
        {
          node: { foo: { bar: 'bazz' } },
          edge: {},
        },
        testing.deleteType('node', 'foo'),
      );

      expect(result).toEqual({
        node: {},
        edge: {},
      });
    });

    it('CREATE_VARIABLE', () => {
      const result = reducer(
        {
          node: { foo: { variables: {} } },
          edge: {},
        },
        testing.createVariable('node', 'foo', 'bar', { baz: 'buzz' }),
      );

      expect(result).toEqual({
        node: { foo: { variables: { bar: { baz: 'buzz' } } } },
        edge: {},
      });
    });

    it('UPDATE_VARIABLE', () => {
      const result = reducer(
        {
          node: { foo: { variables: { bar: { baz: 'buzz' } } } },
          edge: {},
        },
        actionCreators.updateVariable('node', 'foo', 'bar', { fizz: 'pop' }),
      );

      expect(result).toEqual({
        node: { foo: { variables: { bar: { fizz: 'pop' } } } },
        edge: {},
      });
    });
  });

  describe('async actions', () => {
    describe('createType()', () => {
      it('dispatches the CREATE_TYPE action with a type id', () => {
        const store = mockStore(mockState);

        store.dispatch(actionCreators.createType(
          'node',
          { fizz: 'buzz' },
        ));

        const actions = store.getActions();

        expect(actions[0]).toMatchObject({
          type: actionTypes.CREATE_TYPE,
          meta: {
            entity: 'node',
            type: uuid(),
          },
          configuration: { fizz: 'buzz' },
        });
      });
    });

    describe('createVariable()', () => {
      it('dispatches the CREATE_VARIABLE action with a variable id', () => {
        const store = mockStore(mockState);

        store.dispatch(actionCreators.createVariable(
          'node',
          'foo',
          { fizz: 'buzz' },
        ));

        const actions = store.getActions();

        expect(actions[0]).toMatchObject({
          type: actionTypes.CREATE_VARIABLE,
          meta: {
            entity: 'node',
            type: 'foo',
            variable: uuid(),
          },
          configuration: { fizz: 'buzz' },
        });
      });
    });

    describe('deleteType()', () => {
      it('Dispatches delete actions for all related objects', () => {
        const mockStateWithProtocol = {
          protocol: {
            present: {
              stages: [
                {
                  id: 'bazz',
                  subject: { entity: 'node', type: 'foo' },
                },
                {
                  id: 'buzz',
                  prompts: [
                    {
                      id: 'fizz',
                      subject: { entity: 'node', type: 'foo' },
                    },
                  ],
                },
              ],
              forms: {
                bar: {
                  entity: 'node',
                  type: 'foo',
                },
              },
            },
          },
        };

        const store = mockStore(mockStateWithProtocol);

        store.dispatch(actionCreators.deleteType('node', 'foo', true));

        const actions = store.getActions();

        const expectedActions = sortByType([
          testing.deleteType('node', 'foo'),
          stageActions.deleteStage('bazz'),
          stageActions.deletePrompt('buzz', 'fizz', true),
          formActions.deleteForm('bar'),
        ]);

        expect(sortByType(actions)).toEqual(expectedActions);
      });
    });
  });
});
