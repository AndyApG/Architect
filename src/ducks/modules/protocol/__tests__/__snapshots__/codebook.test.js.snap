// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`protocol.codebook DELETE_VARIABLE DELETE_VARIABLE for node entity 1`] = `
Object {
  "edge": Object {},
  "node": Object {
    "foo": Object {
      "variables": Object {},
    },
  },
}
`;

exports[`protocol.codebook async actions createVariable() dispatches the CREATE_VARIABLE action with a variable id for ego 1`] = `
Object {
  "configuration": Object {
    "fizz": "buzz",
    "name": "bar",
    "type": "text",
  },
  "meta": Object {
    "entity": "ego",
    "type": undefined,
    "variable": "809895df-bbd7-4c76-ac58-e6ada2625f9b",
  },
  "type": "PROTOCOL/CREATE_VARIABLE",
}
`;

exports[`protocol.codebook async actions createVariable() dispatches the CREATE_VARIABLE action with a variable id for node 1`] = `
Object {
  "configuration": Object {
    "fizz": "buzz",
    "name": "bar",
    "type": "text",
  },
  "meta": Object {
    "entity": "node",
    "type": "foo",
    "variable": "809895df-bbd7-4c76-ac58-e6ada2625f9b",
  },
  "type": "PROTOCOL/CREATE_VARIABLE",
}
`;

exports[`protocol.codebook reducer CREATE_VARIABLE CREATE_VARIABLE for ego entity 1`] = `
Object {
  "ego": Object {
    "variables": Object {
      "bar": Object {
        "baz": "buzz",
      },
    },
  },
}
`;

exports[`protocol.codebook reducer CREATE_VARIABLE CREATE_VARIABLE for node entity 1`] = `
Object {
  "edge": Object {},
  "node": Object {
    "foo": Object {
      "variables": Object {
        "bar": Object {
          "baz": "buzz",
        },
      },
    },
  },
}
`;

exports[`protocol.codebook reducer UPDATE_VARIABLE UPDATE_VARIABLE for ego entity 1`] = `
Object {
  "ego": Object {
    "variables": Object {
      "bar": Object {
        "fizz": "pop",
      },
    },
  },
}
`;

exports[`protocol.codebook reducer UPDATE_VARIABLE UPDATE_VARIABLE for node entity 1`] = `
Object {
  "edge": Object {},
  "node": Object {
    "foo": Object {
      "variables": Object {
        "bar": Object {
          "fizz": "pop",
        },
      },
    },
  },
}
`;
